jni/src/faiss_wrapper.cpp

```diff
@@ -0,0 +1,357 @@
+    std::string indexDescriptionCpp(knn_jni::ConvertJavaObjectToCppString(env, indexDescriptionJ));
+
+    std::unique_ptr<faiss::Index> indexWriter;
+    indexWriter.reset(faiss::index_factory(dim, indexDescriptionCpp.c_str(), metric));
+
+    // Add extra parameters that cant be configured with the index factory
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        jobject subParametersJ = parametersCpp[knn_jni::PARAMETERS];
+        auto subParametersCpp = knn_jni::ConvertJavaMapToCppMap(env, subParametersJ);
+        SetExtraParameters(env, subParametersCpp, indexWriter.get());
+        env->DeleteLocalRef(subParametersJ);
+    }
+
+    // Train index if needed -- check if there is a case where index needs part trained but is trained
+    if(!indexWriter->is_trained) {
+        //TODO: What needs to be freed???
+        throw std::runtime_error("Index is not trained");
+    }
+    env->DeleteLocalRef(parametersJ);
+
+    auto idVector = knn_jni::ConvertJavaIntArrayToCppIntVector(env, idsJ);
+    faiss::IndexIDMap idMap =  faiss::IndexIDMap(indexWriter.get());
+    idMap.add_with_ids(numVectors, dataset.data(), idVector.data());
+
+    // Write the index to disk
+    std::string indexPathCpp(ConvertJavaStringToCppString(env, indexPathJ));
+    faiss::write_index(&idMap, indexPathCpp.c_str());
+}
+
+void knn_jni::faiss_wrapper::createIndexFromTemplate(JNIEnv * env, jintArray idsJ, jobjectArray vectorsJ,
+                                                     jstring indexPathJ, jbyteArray templateIndexJ) {
+    try {
+        if (idsJ == nullptr) {
+            throw std::runtime_error("IDs cannot be null");
+        }
+
+        if (vectorsJ == nullptr) {
+            throw std::runtime_error("Vectors cannot be null");
+        }
+
+        if (indexPathJ == nullptr) {
+            throw std::runtime_error("Index path cannot be null");
+        }
+
+        if (templateIndexJ == nullptr) {
+            throw std::runtime_error("Parameters cannot be null");
+        }
+    } catch (...) {
+        CatchCppExceptionAndThrowJava(env);
+    }
+
+    // Read data set
+    int numVectors = knn_jni::GetJavaObjectArrayLength(env, vectorsJ);
+    int numIds = knn_jni::GetJavaIntArrayLength(env, idsJ);
+    if (numIds != numVectors) {
+        throw std::runtime_error("Number of IDs does not match number of vectors");
+    }
+
+    int dim = knn_jni::GetInnerDimensionOf2dJavaArray(env, vectorsJ);
+    auto dataset = knn_jni::Convert2dJavaObjectArrayToCppFloatVector(env, vectorsJ, dim);
+
+    // Get vector of bytes from jbytearray
+    int indexBytesCount = knn_jni::GetJavaBytesArrayLength(env, templateIndexJ);
+    jbyte * indexBytesJ = env->GetByteArrayElements(templateIndexJ, nullptr);
+    faiss::VectorIOReader vectorIoReader;
+    for (int i = 0; i < indexBytesCount; i++) {
+        vectorIoReader.data.push_back((uint8_t) indexBytesJ[i]);
+    }
+    env->ReleaseByteArrayElements(templateIndexJ, indexBytesJ, 0);
```

Comment:
Leaks if any of the above fails? `indexBytesJ` needs to be inside a smart pointer that calls `ReleaseByteArrayElements` when it's deleted.

Reply:
Will fix.

Reply:
For this, I am not sure smart pointer is necessary. `ReleaseByteArrayElements` requires the Java object to be passed as the first argument. So, we would need to construct a wrapper that includes a reference to the Java object. Because the only code between allocation and release is appending to a vector, I think it is okay to keep this as a raw pointer.